File: problems/222-count-complete-tree-nodes/dfs-pre-order.cpp
/**
 * DFS
 * 
 * 时间：`O(N)`, 44ms
 */

class Solution {
public:
    int countNodes(TreeNode* root) {
        res = 0;
        dfs(root);
        return res;
    }

    void dfs(TreeNode* node) {
        if (!node) return;

        ++res;
        dfs(node->left);
        dfs(node->right);
    }

private:
    int res;
};

File: 剑指offer/27/dfs-pre-order.js
/**
 * 前序遍历
 */
var mirrorTree = function (root) {
  if (!root) return root;

  // 先反转本结点  
  [root.left, root.right] = [root.right, root.left];
  // 再递归子树
  root.left = mirrorTree(root.left);
  root.right = mirrorTree(root.right);

  return root;
};